### PR TITLE
Add an ExecuteLastObjectDataManager that attempts to execute after ot…

### PR DESCRIPTION
**Pull request by jzuech3**
…her ObjectDataManagers #10 @papachoco @jamadden

### CODE REVIEW

File: src/nti/transactions/transactions.py [hunk 1]
```diff
@@ -237,6 +253,16 @@ def do(*args, **kwargs):
     transaction.get().join(result)
     return result
 
+def do_last(*args, **kwargs):
```
Comment (jamadden) [inline src/nti/transactions/transactions.py]: Can we not use the word "last", here or in the class? That implies promises we can't keep, and it makes no sense if this function/class is used more than once.

Comment (jzuech3) [reply src/nti/transactions/transactions.py]: Recommendations? I'm stumped :)

File: src/nti/transactions/transactions.py [hunk 2]
```diff
@@ -237,6 +253,16 @@ def do(*args, **kwargs):
...
+    """
+    Establishes a IDataManager in the current transaction that will attempt to
+    execute *after* all other DataManagers have had their say.
+    See :class:`ObjectDataManager` for the possible arguments.
+    """
+    result = ExecuteLastObjectDataManager(*args, **kwargs)
```
Comment (jamadden) [inline src/nti/transactions/transactions.py]: Rather than reimplementing this, what about something like this:

```python
def do(*args, **kwargs):
    klass = kwargs.pop('datamanager_class', ObjectDataManager) # new argument
    result = klass(*args, **kwargs)
    ... # as before

def do_XXX(*args, **kwargs):
   kwargs['datamanager_class'] = XXXDataManager
   return do(*args, **kwargs)
```

File: src/nti/transactions/transactions.py (hunk 1 above)
Comment (jamadden) [reply src/nti/transactions/transactions.py]: `OrderedNearEndObjectDataManager` and `do_near_end`? (I suppose even `NearLast` would get the point across.)

Comment (jzuech3) [reply src/nti/transactions/transactions.py]: I'm fine with either of those. I was hoping the comments would provide the caveat.

**Comment by cutz** [comment]
out of curiosity how does this work with the single phase commit transactions in the zope transaction manager? IIRC those get sorted to the end by the transaction manager (i've seen this in certain implementations of sqlalchemy).  Does this sort before or after those?

File: src/nti/transactions/transactions.py (hunk 1 above)
Comment (jamadden) [reply src/nti/transactions/transactions.py]: Only if you read the docstrings. That doesn't help someone just reading the code where this is used.

**Comment by jamadden** [comment]
The zope transaction manager doesn't have a "single phase commit transaction" so I'm not sure what you mean. It only supports two-phase commit. Individual data managers may or may not do something useful during all of those phases.

**Comment by cutz** [comment]
Sorry it was the specific sqlalchemy transaction datamanager I was remembering. It tries to sort single phase committers to the end. 
https://github.com/zopefoundation/zope.sqlalchemy/blob/master/src/zope/sqlalchemy/datamanager.py#L118

**Comment by jamadden** [comment]
`~` sorts after `zzz` so sqlalchemy will still go last. (Which is a good example of why these shouldn't be named `Last` 😄 )

```
$ python -c 'print(sorted(["~sqlalchemy:123", "zzz123"]))'
['zzz123', '~sqlalchemy:123']
```

File: src/nti/transactions/tests/test_transaction.py [hunk 3]
```diff
@@ -12,17 +12,22 @@
...
 from hamcrest import raises
+from hamcrest import contains
 
 import fudge
 
-from ..transactions import _do_commit
-from ..transactions import TransactionLoop
-from ..interfaces import CommitFailedError
-from ..interfaces import AbortFailedError
+from nti.transactions.interfaces import CommitFailedError
```
Comment (jamadden) [inline src/nti/transactions/tests/test_transaction.py]: Why this change?

File: src/nti/transactions/tests/test_transaction.py [hunk 4]
```diff
@@ -218,3 +223,19 @@ def handler():
...
+            results.append(x)
+
+        # The managers will execute in order added (since identical),
+        # except for the one that requests to go last.
+        manager1 = do(call=test_call, args=(0,))
+        manager2 = do(call=test_call, args=(1,))
+        manager_post = do_near_end(call=test_call, args=(10,))
+        manager3 = do(call=test_call, args=(2,))
+        transaction.commit()
+        assert_that(results, contains(0,1,2,10))
```
Comment (jamadden) [inline src/nti/transactions/tests/test_transaction.py]: Spaces between commas, please.

**Review by jamadden — APPROVED** [review]
Some minor formatting issues.

File: src/nti/transactions/tests/test_transaction.py (hunk 3 above)
Comment (jzuech3) [reply src/nti/transactions/tests/test_transaction.py]: I thought we preferred absolute imports?

Comment (jamadden) [reply src/nti/transactions/tests/test_transaction.py]: I don't, they make refactoring harder.

**Review by jamadden — CHANGES_REQUESTED** [review]
Can you add a note to CHANGES.rst? And add `..versionadded:: 1.1` lines to the end of the docstrings for the new objects?

Sorry, I should have mentioned that earlier.

**Review by jamadden — APPROVED** [review]
